README
NAME
    Win32::CommandLine - Retrieve and reparse the Win32 command line

VERSION
    our $VERSION = qw$Version: 0.5.8.124 $[1]

SYNOPSIS
            @ARGV = Win32::CommandLine::argv() if eval { require Win32::CommandLine; };

            _or_

            use Win32::CommandLine qw( command_line );
            my $commandline = command_line();
            ...

DESCRIPTION
    This module is used to reparse the Win32 command line, automating better
    quoting and globbing of the command line. Globbing is full bash POSIX
    compatible globbing, including subshell expansions. With the use of the
    companion script (`xx.bat') and `doskey' for macro aliasing, you can add
    fully implemented, bash compatible, string quoting/expansion and file
    globbing to *any* Win32 executable.

    This module is compatible with both `cmd.exe' and `4nt/tcc/tcmd' shells,
    and can be used to add better parsing and bash glob expansion to *any*
    external command (by using the included `xx.bat' batch script).

  `CMD.EXE'
        doskey type=call xx type $*
        type [a-c]*.pl
        doskey perl=call xx perl $*
        perl -e 'print "test"'              [would otherwise FAIL]

  `4NT/TCC/TCMD'
        alias type=call xx type
        type [a-c]*.pl
        .
        alias perl=call xx perl
        perl -e 'print "test"'              [would otherwise FAIL]

    Note that bash compatible character expansion and globbing is available,
    including glob meta-notations such as "`a[bc]*'" or
    "`foo.{bat,pl,exe,o}'".

  Command line string/character expansion
      '...'    literal (no escapes and no globbing within quotes)
      "..."    literal (no escapes and no globbing within quotes)
      $'...'   string including all ANSI C string escapes (see NOTE); no globbing within quotes
      $"..."   literal (no escapes and no globbing within quotes) [same as "..."]
      $( ... ) subshell expansion [subshell commandline is _not_ automatically expanded]
      $("...") subshell expansion (quotes removed) [subshell commandline is _not_ automatically expanded]

    NOTE: ANSI C string escapes are (\a, \b, \e, \f, \n, \r, \t, \v, \\, \',
    \[0-9]{1,3}, \x[0-9a-fA-f]{1,2}, \c[@A-Z[\\\]\^_`]; all other escaped
    characters are left in place without transformation (\<x> => \<x>).

    URLref: [bash ANSI-C Quoting]
    http://www.gnu.org/software/bash/manual/html_node/ANSI_002dC-Quoting.htm
    l @@ http://www.webcitation.org/66M8skmP8

  GLOB META CHARACTERS
      \           Quote the next metacharacter
      []          Character class
      {}          Multiple pattern
      *           Match any string of characters
      ?           Match any single character
      ~           Current user home directory
      ~USERNAME   Home directory of USERNAME
      ~TEXT       Environment variable named ~TEXT (aka $ENV{~TEXT}) [overrides ~USERNAME expansion]

    The multiple pattern metanotation `a{b,c,d}e' is a shorthand for `abe
    ace ade'. Left to right order is preserved, with results of matches
    being sorted separately at a low level to preserve this order.

INSTALLATION
    To install this module, run the following commands:

        perl Build.PL
        ./Build
        ./Build test
        ./Build install

    Or, if you're on a platform (like DOS or Windows) that doesn't require
    the "./" notation, you can do this:

        perl Build.PL
        Build
        Build test
        Build install

    Alternatively, the standard make idiom is also available (though it is
    deprecated):

        perl Makefile.PL
        make
        make test
        make install

    (On Windows platforms you should use `nmake' or `dmake' instead.)

    Note that the Makefile.PL script is just a pass-through, and
    Module::Build is still ultimately required for installation. Makefile.PL
    will throw an exception if Module::Build is missing from your current
    installation. `cpan' will notify the user of the build prerequisites
    (and install them for the build, if it is setup to do so [see the cpan
    configuration option `build_requires_install_policy']).

    PPM installation bundles should also be available in the standard PPM
    repositories (i.e. ActiveState, trouchelle.com
    [http://trouchelle.com/ppm/package.xml]).

    Note: On ActivePerl installations, "`./Build install'" will do a full
    installation using `ppm' (see ppm). During the installation, a PPM
    package is constructed locally and then subsequently used for the final
    module install. This allows for uninstalls (using "`ppm uninstall
    '*`MODULE'*" and also keeps local HTML documentation current.

RATIONALE
    This began as a simple need to work-around the less-than-stellar
    `COMMAND.COM'/`CMD.EXE' command line parser, just to accomplish more
    `correct` quotation interpretation. It then grew into a small odyssey:
    learning XS and how to create a perl module, learning the perl build
    process and creating a customized build script/environment, researching
    tools and developing methods for revision control and versioning,
    learning and creating perl testing processes, and finally learning about
    PAUSE and perl publishing practices. And, somewhere in the middle,
    adding some of the `bash' shell magic to the CMD shell (and,
    additionally, making it compatible with the excellent [and free] TCC-LE
    shell from JPSoft [find that at http://jpsoft.com]).

    Some initial attempts were made using `Win32::API' and `Inline::C'. For
    example (`Win32::API' attempt [caused GPFs]):

      @rem = '--*-Perl-*--
      @echo off
      if "%OS%" == "Windows_NT" goto WinNT
      perl -x -S "%0" %1 %2 %3 %4 %5 %6 %7 %8 %9
      goto endofperl
      :WinNT
      perl -x -S %0 %*
      if NOT "%COMSPEC%" == "%SystemRoot%\system32\cmd.exe" goto endofperl
      if %errorlevel% == 9009 echo You do not have Perl in your PATH.
      if errorlevel 1 goto script_failed_so_exit_with_non_zero_val 2>nul
      goto endofperl
      @rem ';
      #!/usr/bin/perl -w
      #line 15
      #
      use Win32::API;
      #
      Win32::API->Import("kernel32", "LPTSTR GetCommandLine()");
      my $string = pack("Z*", GetCommandLine());
      #
      print "string[".length($string)."] = '$string'\n";
      # ------ padding --------------------------------------------------------------------------------------
      __END__
      :endofperl

    Unfortunately, `Win32::API' and `Inline::C' were shown to be too fragile
    at the time (in 2007). `Win32::API' caused occasional (but reproducible)
    GPFs, and `Inline::C' was very brittle on Win32 systems (not
    compensating for paths with embedded strings). [ See
    http://www.perlmonks.org/?node_id=625182 for a more full explanation of
    the problem and initial attempts at a solution. ]

    So, an initial XS solution was implemented. And from that point, the
    lure of `bash'-like command line parsing led inexorably to the full
    implementation. The parsing logic is unfortunately still complex, but
    seems to be holding up under testing.

DEPENDENCIES
    `Win32::CommandLine' requires `Carp::Assert' for internal error checking
    and warnings.

    The optional modules `Win32', `Win32::Security::SID', and
    `Win32::TieRegistry' are recommended to allow full glob tilde expansions
    for user home directories (eg, `~administrator' expands to
    `C:\Users\Administrator'). Expansion of the single tilde (`~') has a
    backup implementation based on %ENV variables, and therefore will still
    work even without the optional modules.

LICENSE AND COPYRIGHT
      Copyright (c) 2007-2009, Roy Ivy III <rivy[at]cpan[dot]org>. All rights reserved.

    This module is free software; you can redistribute it and/or modify it
    under the Perl Artistic License v2.0 (see
    http://opensource.org/licenses/artistic-license-2.0.php).

DISCLAIMER OF WARRANTY
      THIS PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER AND CONTRIBUTORS "AS IS"
      AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED WARRANTIES OF
      MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, OR NON-INFRINGEMENT
      ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS REQUIRED
      BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
      INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF
      THE USE OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

      [REFER TO THE FULL LICENSE FOR EXPLICIT DEFINITIONS OF ALL TERMS.]

ACKNOWLEDGEMENTS
    Thanks to BrowserUK and syphilis (aka SISYPHUS on CPAN) for some helpful
    ideas (including an initial XS starting point for the module) during a
    discussion on PerlMonks (see http://www.perlmonks.org/?node_id=625182).

AUTHOR
    Roy Ivy III <rivy[at]cpan[dot]org>



Changes

Revision history for Win32::CommandLine

0.01 Sun Jul  8 19:46:23 2007
    - Initial release.  Created by ExtUtils::ModuleMaker 0.49.




t/prelim/echo.cs
using System;

namespace APP { class APP {

public static void Main(string[] args)
{
	string output = "";
	for (int i = 0; i < args.Length; i++) { output += "[" + i + "]" + args[i] + "\n"; }
	Console.WriteLine( output );
}

}}

t/prelim/echo.pl
#!perl -w   -*- tab-width: 4; mode: perl -*-
use strict; use diagnostics;

use Win32::CommandLine;

print "command_line() = `".Win32::CommandLine::command_line()."`\n";
#print "-- ORIGINAL\n";
#print join(" ",@ARGV)."\n";

#@ARGV = Win32::CommandLine::argv() if eval { require Win32::CommandLine; };
#print "argv() = {@ARGV}\n";

@ARGV = Win32::CommandLine::_argv( Win32::CommandLine::command_line() ) if eval { require Win32::CommandLine; };
print "_argv() = {@ARGV}\n";

@ARGV = Win32::CommandLine::_argv( Win32::CommandLine::command_line(), {dosify => 1}  ) if eval { require Win32::CommandLine; };
print "_argv(dosify=t) = {@ARGV}\n";

#@ARGV = Win32::CommandLine::_argv_NEW( Win32::CommandLine::command_line() ) if eval { require Win32::CommandLine; };
#print "_argvNEW() = {@ARGV}\n";
#@ARGV = Win32::CommandLine::_argv_NEW( Win32::CommandLine::command_line(), {dosify => 1} ) if eval { require Win32::CommandLine; };
#print "_argvNEW(dosify=t) = {@ARGV}\n";

#print "-- REINTERPRETED\n";

#print "-- COMMANDLINE\n";
#print Win32::CommandLine::command_line()."\n";

#print "-- ENV{CMDLINE}\n";
#print $ENV{CMDLINE}."\n";


extras/digest-md5.pl
#!perl -w   -- -*- tab-width: 4; mode: perl -*-
#$Id: digest-md5.pl,v 0.5.8.124 ( r195:9bba78c5cd5e [mercurial] ) 2009/04/29 05:27:31 rivy $

# Script Summary

=head1 NAME

digest-md5 - Find and print the executable path(s)

=head1 VERSION

This document describes C<digest-md5> ($Version: 0.5.8.124 $).

=head1 SYNOPSIS

digest-md5 [B<<option(s)>>] B<<filename(s)>>

=begin HIDDEN-OPTIONS

Options:

		--version       version message
	-?, --help          brief help message

=end HIDDEN-OPTIONS

=head1 OPTIONS

=over

=item --version

=item --usage

=item --help, -?

=item --man

Print the usual program information

=back

=head1 REQUIRED ARGUMENTS

=over

=item <filename(s)>

FILENAMES for digest.

=back

=head1 DESCRIPTION

B<digest-md5> will calculate and print the hexadecimal MD5 digest for each FILENAME.

=cut

# VERSION: major.minor.release[.build]]  { minor is ODD => alpha/beta/experimental; minor is EVEN => stable/release }
# generate VERSION from $Version: 0.5.8.124 $ SCS tag
# $defaultVERSION 	:: used to make the VERSION code resilient vs missing keyword expansion
# $generate_alphas	:: 0 => generate normal versions; true/non-0 => generate alpha version strings for ODD numbered minor versions
use version qw(); our $VERSION; { my $defaultVERSION = '0.1.0'; my $generate_alphas = 0; $VERSION = ( $defaultVERSION, qw( $Version: 0.5.8.124 $ ))[-2]; if ($generate_alphas) { $VERSION =~ /(\d+)\.(\d+)\.(\d+)(?:\.)?(.*)/; $VERSION = $1.'.'.$2.((!$4&&($2%2))?'_':'.').$3.($4?((($2%2)?'_':'.').$4):q{}); $VERSION = version::qv( $VERSION ); }; } ## no critic ( ProhibitCallsToUnexportedSubs ProhibitCaptureWithoutTest ProhibitNoisyQuotes ProhibitMixedCaseVars ProhibitMagicNumbers)

use Pod::Usage;
use Getopt::Long qw(:config bundling bundling_override gnu_compat no_getopt_compat);

#use Carp::Assert;

use strict;
use warnings;
use diagnostics;

use File::Spec;

use Digest::MD5;

@ARGV = Win32::CommandLine::argv() if eval { require Win32::CommandLine; };

# getopt
my %ARGV = ();
GetOptions (\%ARGV, 'help|h|?|usage', 'man', 'version|ver|v') or pod2usage(2);
#Getopt::Long::VersionMessage() if $ARGV{'version'};
pod2usage(-verbose => 99, -sections => '', -message => (File::Spec->splitpath($0))[2]." v$::VERSION") if $ARGV{'version'};
pod2usage(1) if $ARGV{'help'};
pod2usage(-verbose => 2) if $ARGV{'man'};

pod2usage(1) if @ARGV < 1;

foreach (@ARGV)
	{
	#print '#args = '.scalar(@ARGV)."\n";
	if (@ARGV > 1) { print "$_: "; }

	open(FILE, $_) or die "Can't open '$_': $!";
    binmode(FILE);

    print Digest::MD5->new->addfile(*FILE)->hexdigest, "\n";

    close(FILE);
	}


extras/which.pl
#!perl -w   -- -*- tab-width: 4; mode: perl -*-
#$Id: which.pl,v 0.5.8.124 ( r227:3490c923d48e [mercurial] ) 2009/06/14 18:18:05 rivy $

## TODO: aliases? bash which doesn't see aliases -- make a switch to search aliases as well?

# Script Summary

=head1 NAME

which - Find and print the executable path(s)

=head1 VERSION

This document describes C<which> ($Version: 0.5.8.124 $).

=head1 SYNOPSIS

which [B<<option(s)>>] B<<filename(s)>>

=begin HIDDEN-OPTIONS

Options:

		--version       version message
	-?, --help          brief help message

=end HIDDEN-OPTIONS

=head1 OPTIONS

=over

=item --dosify, --dos, --winify, --win

"Dosify" output ( slashed/quoted appropriately for CMD or TCC )

=item TODO --msysify, --msys, --cygify, --cygwin

"MSYS"ify output ( convert to '/c/...' )

=item TODO --unixify, --unix

"Unixify" output ( slashed/quoted appropriately for SH or BASH )

=item TODO --auto

Choose --dos, --unix, or --msys depending on surrounding execution context

=item TODO --quoting-style=QUOTETYPE

Quoting style = QUOTETYPE (literal, shell, shell-always, c, escape) [ see LS.c and quoteargs.c code; see URLrefs: http://www.ss64.com/bash/ls.html ; http://linux.about.com/od/commands/l/blcmdl1_patch.htm ; http://www.gnu.org/software/coreutils/manual/html_node/What-information-is-listed.html ]
[ modified by $ENV{QUOTING_STYLE} ]

=item TODO --shell=SHELLTYPE

Shell type = SHELLTYPE (auto [DEFAULT], dos / cmd / win / win32 / windows, msys / cygwin, unix / bash / sh )
[ modified by $ENV{QUOTING_STYLE} ]

=item --where, -w, --all, -a

Find and print B<<all>> possible executable paths for the filename(s) given

=item --version

=item --usage

=item --help, -?

=item --man

Print the usual program information

=back

=head1 REQUIRED ARGUMENTS

=over

=item <filename(s)>

FILENAMES...

=back

=head1 DESCRIPTION

B<which> will read each FILENAME, find, and then print the executable path for the FILENAME.

=cut

# VERSION: major.minor.release[.build]]  { minor is ODD => alpha/beta/experimental; minor is EVEN => stable/release }
# generate VERSION from $Version: 0.5.8.124 $ SCS tag
# $defaultVERSION 	:: used to make the VERSION code resilient vs missing keyword expansion
# $generate_alphas	:: 0 => generate normal versions; true/non-0 => generate alpha version strings for ODD numbered minor versions
use version qw(); our $VERSION; { my $defaultVERSION = '0.1.0'; my $generate_alphas = 0; $VERSION = ( $defaultVERSION, qw( $Version: 0.5.8.124 $ ))[-2]; if ($generate_alphas) { $VERSION =~ /(\d+)\.(\d+)\.(\d+)(?:\.)?(.*)/; $VERSION = $1.'.'.$2.((!$4&&($2%2))?'_':'.').$3.($4?((($2%2)?'_':'.').$4):q{}); $VERSION = version::qv( $VERSION ); }; } ## no critic ( ProhibitCallsToUnexportedSubs ProhibitCaptureWithoutTest ProhibitNoisyQuotes ProhibitMixedCaseVars ProhibitMagicNumbers)

use Pod::Usage;
use Getopt::Long qw(:config bundling bundling_override gnu_compat no_getopt_compat);

#use Carp::Assert;

use strict;
use warnings;
use diagnostics;

#use File::Which;
use File::Spec;

use Env::Path qw(PATH);

#sub dosify;

@ARGV = Win32::CommandLine::argv( {glob => 0} ) if eval { require Win32::CommandLine; };

# getopt
my %ARGV = ();
GetOptions (\%ARGV, 'where|w|all|a', 'help|h|?|usage', 'man', 'version|ver|v', 'dosify|dos|d') or pod2usage(2);
#Getopt::Long::VersionMessage() if $ARGV{'version'};
pod2usage(-verbose => 99, -sections => '', -message => (File::Spec->splitpath($0))[2]." v$::VERSION") if $ARGV{'version'};
pod2usage(1) if $ARGV{'help'};
pod2usage(-verbose => 2) if $ARGV{'man'};

pod2usage(1) if @ARGV < 1;

if ($^O eq "MSWin32") { PATH->Prepend( q{.} ); }

PATH->Uniqify();

$ARGV{dosify} = 1;		## TODO: add options to avoid hard-coding this

foreach (@ARGV)
	{
	#print '#args = '.scalar(@ARGV)."\n";
	##if (@ARGV > 1) { print "$_:\n"; }
#	my @w = ($ARGV{where} ? which( $_ ) : scalar(which( $_ )) );

	# TODO: aliases
	# how to note these? include entire text/options? what about aliases with multiple commands?
	# CMD = 'doskey /macros'
	# perl -e "$o = qx{doskey /macros}; print $o;"
	# [format]
    # cpan=wrap-cpan $*
    # command=cmd $*
    # to-hex=perl -e "printf q{%x},$1"
    # vdiff=winmerge $*
    # bt=b test --test_files $*
    # b=ant+build $*
    # ant="c:/program files/ant/bin/ant" $*
    # oh=handle $*
    # morehelp=hh ntcmds.chm $*
    # cdg=x -S cd $*
    # lag=x msls -lA $*
    # llg=x msls -l $*
    # lsg=x msls $*
    # dirg=x msls $*
    # la=msls -lA $*
    # ll=msls -l $*
    # ls=msls $*
    # dir=msls -l $*
    # args=xx -a $*
    # perl=call xx perl $*
    # pdh=start c:\perl\html\index.html $*
    # pd=perldoc $*
    # fs=findstr /s /i $*
    # ds=dir /s /b $*
    # e=edit $*
    # .....=cd ..\..\..\.. $*
    # ....=cd ..\..\.. $*
    # ...=cd ..\.. $*
    # ..=cd .. $*
    # whois=pwhois -s -S -c -C $*
    # rn=rename $*
    # mv=move $*
    # rm=erase $*
    # cp=copy $*
    # NOTE: trailing $* is optional; $T is a command seperator (multiple commands are possible); $1-$9 batch parameters are possible

	# TCC = 'alias'
	# perl -e "$o = qx{tcc /c alias}; print $o;"
	# [format]
	# cp=copy
	# rm=erase
	# mv=move
	# rn=rename
	# whois=pwhois -s -S -c -C
	# ..=cd ..
	# ...=cd ..\..
	# ....=cd ..\..\..
	# .....=cd ..\..\..\..
	# e=edit
	# ds=dir /s /b
	# fs=findstr /s /i
	# pd=perldoc
	# pdh=start c:\perl\html\index.html
	# perl=call xx perl
	# args=xx -a
	# dir=msls -l
	# ls=msls
	# ll=msls -l
	# la=msls -lA
	# dirg=x msls
	# lsg=x msls
	# llg=x msls -l
	# lag=x msls -lA
	# cdg=x -S cd
	# morehelp=hh ntcmds.chm
	# oh=handle
	# ant="c:/program files/ant/bin/ant"
	# b=*ant+build
	# bt=b test --test_files
	# vdiff=winmerge
	# to-hex=perl -e "printf q{%%x},%1"
	# command=cmd
	# cpan=wrap-cpan

	# bash = 'alias'
	# perl -e "\$o = qx{/bin/bash -ic alias}; print \$o;"
	# NOTE: this only finds aliases defined in .bashrc (not the current shell)
	# [format]
	# alias ant='c:/program\ files/ant/bin/ant'
	# alias b='c:/users/public/documents/\@bin/ant+build.pl'
	# alias bt='b test --test_files'
	# alias df='df -h'
	# alias dir='ls'
	# alias du='du -h'
	# alias la='ls -A'
	# alias ll='ls -l'
	# alias ls='ls --color=auto -hF'
	# alias mc='. /usr/share/mc/bin/mc-wrapper.sh'
	# alias pd='perldoc'
	# NOTE: mc is aliased to a sourced shell script [? return '.', '. /usr/share/mc/bin/mc-wrapper.sh', or '/usr/share/mc/bin/mc-wrapper.sh']

	my @w = PATH->Whence( $_ );
	if (! $ARGV{where} && @w) { @w = $w[0]; }
	my %printed;
#	# output full path for all matches (and no repeats [repeats can happen if the PATH contains multiple references to the same location])
	for (@w) { if ($_) {$_ = File::Spec->rel2abs($_); if (!$printed{$_}) {$printed{$_}=1; print ''.($ARGV{dosify} ? _dosify_path($_) : $_)."\n"; } } }
	#if (@w) { print join("\n", @w)."\n"; }
	}

sub _dosify_path {
	# _dosify_path( <null>|$|@ ): returns <null>|$|@ ['shortcut' function]
	# dosify string (assumed to be a path), returning a string which will be interpreted/parsed by DOS/CMD as the input path when parsed at the command line
	@_ = @_ ? @_ : $_ if defined wantarray;		## no critic (ProhibitPostfixControls)	## break aliasing if non-void return context

	for (@_ ? @_ : $_)
		{
		s:\/:\\:g;								# forward to back slashes
		_dosify($_);
		}

	return wantarray ? @_ : "@_";
	}

sub	_dosify {
	# _dosify( <null>|$|@ ): returns <null>|$|@ ['shortcut' function]
	# dosify string, returning a string which will be interpreted/parsed by DOS/CMD as the input string when input to the command line
	# CMD/DOS quirks: dosify double-quotes:: {\\} => {\\} UNLESS followed by a double-quote mark when {\\} => {\} and {\"} => {"} (and doesn't end the quote)
	#	:: EXAMPLES: {a"b"c d} => {[abc][d]}, {a"\b"c d} => {[a\bc][d]}, {a"\b\"c d} => {[a\b"c d]}, {a"\b\"c" d} => {[a\b"c"][d]}
	#				 {a"\b\\"c d} => {[a\b\c][d]}, {a"\b\\"c" d} => {[a\b\c d]}, {a"\b\\"c d} => {[a\b\c][d]}, {a"\b\\c d} => {[a\b\\c d]}
	@_ = @_ ? @_ : $_ if defined wantarray;		## no critic (ProhibitPostfixControls)	## break aliasing if non-void return context

	# TODO: check these characters for necessity => PIPE characters [<>|] and internal double quotes for sure, _likely_ escape character [^], [:]?, [*?] glob chars needed?, what about glob character set chars [{}]?
	my $dos_special_chars = '"<>|^';
	my $dc = quotemeta( $dos_special_chars );
	for (@_ ? @_ : $_)
		{
		#print "_ = $_\n";
		#s:\/:\\:g;								# forward to back slashes [now done in _dosify_path()
		if ( $_ =~ qr{(\s|[$dc])} )
			{
			#print "in qr\n";
			s:":\\":g;							# CMD: preserve double-quotes with backslash	# TODO: change to $dos_escape	## no critic (ProhibitUnusualDelimiters)
			s:([\\]+)\\":($1 x 2).q{\\"}:eg;	# double backslashes in front of any \" to preserve them when interpreted by DOS/CMD
			$_ = q{"}.$_.q{"};					# quote the final token
			};
		}

	return wantarray ? @_ : "@_";
}
